shell/browser/file_select_helper.cc

```diff
@@ -0,0 +1,299 @@
+// Copyright (c) 2020 Microsoft, Inc. All rights reserved.
+// Use of this source code is governed by the MIT license that can be
+// found in the LICENSE file.
+
+#include <memory>
+
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "shell/browser/file_select_helper.h"
+
+#include "base/bind.h"
+#include "base/files/file_enumerator.h"
+#include "base/files/file_util.h"
+#include "base/strings/utf_string_conversions.h"
+#include "chrome/common/pref_names.h"
+#include "components/prefs/pref_service.h"
+#include "shell/browser/electron_browser_context.h"
+#include "shell/browser/javascript_environment.h"
+#include "shell/browser/ui/file_dialog.h"
+#include "shell/common/gin_converters/callback_converter.h"
+#include "shell/common/gin_converters/file_path_converter.h"
+
+using blink::mojom::FileChooserFileInfo;
+using blink::mojom::FileChooserFileInfoPtr;
+using blink::mojom::FileChooserParams;
+using blink::mojom::NativeFileInfo;
+
+namespace {
+void DeleteFiles(std::vector<base::FilePath> paths) {
+  for (auto& file_path : paths)
+    base::DeleteFile(file_path);
+}
+}  // namespace
+
+FileSelectHelper::FileSelectHelper(
+    content::RenderFrameHost* render_frame_host,
+    scoped_refptr<content::FileSelectListener> listener,
+    FileChooserParams::Mode mode)
+    : render_frame_host_(render_frame_host),
+      listener_(std::move(listener)),
+      mode_(mode) {
+  DCHECK(render_frame_host_);
+  DCHECK(listener_);
+
+  web_contents_ = content::WebContents::FromRenderFrameHost(render_frame_host);
+  DCHECK(web_contents_);
+  observer_.RemoveAll();
+  content::WebContentsObserver::Observe(web_contents_);
+  observer_.Add(render_frame_host_->GetRenderViewHost()->GetWidget());
+}
+
+FileSelectHelper::~FileSelectHelper() = default;
+
+void FileSelectHelper::ShowOpenDialog(
+    const file_dialog::DialogSettings& settings) {
+  v8::Isolate* isolate = electron::JavascriptEnvironment::GetIsolate();
+  v8::HandleScope scope(isolate);
+  gin_helper::Promise<gin_helper::Dictionary> promise(isolate);
+
+  // Will be released in one of the following
+  // OnFilesSelected - Callback flow completed with response
+  // RunFileChooserEnd - If user cancelled dialog or render_frame_host was
+  // destroyed WebContentsDestroyed
+  AddRefWithCheck();
+  DCHECK(HasAtLeastOneRef());
+
+  auto callback = base::BindOnce(&FileSelectHelper::OnOpenDialogDone,
+                                 base::Unretained(this));
```

Comment:
It's surprising to me that we'd use `base::Unretained` here, given the class is refcounted. It's not clear to me why this class needs to be refcounted, but either we should just past `this` here, or we should change the class to not be refcounted.
```suggestion
                                 this);
```

Reply:
Refcounting is really not needed for this class, for a better DCHECK I was trying to maintain a deterministic refcount. Will rework this.